spring-kafka/src/main/java/org/springframework/kafka/annotation/KafkaListenerAnnotationBeanPostProcessor.java

```diff
@@ -828,7 +829,13 @@ private MessageHandlerMethodFactory createDefaultMessageHandlerMethodFactory() {
 			final GenericMessageConverter messageConverter =
 					new GenericMessageConverter(this.defaultFormattingConversionService);
 			argumentResolvers.add(new MessageMethodArgumentResolver(messageConverter));
+
+			// Custom argument resolution
+			argumentResolvers.addAll(KafkaListenerAnnotationBeanPostProcessor.this.registrar.getCustomMethodArgumentResolvers());
+
+			// Has to be at the end - if not it will try to resolver objects from CustomMethodArgumentResolvers
 			argumentResolvers.add(new KafkaNullAwarePayloadArgumentResolver(messageConverter, validator));
+
 			defaultFactory.setArgumentResolvers(argumentResolvers);
```

Comment:
I think this one has to be changed to `setCustomArgumentResolvers()`.
We have similar logic in Spring Integration and the proper order is applied:https://github.com/spring-projects/spring-integration/blob/master/spring-integration-core/src/main/java/org/springframework/integration/handler/support/MessagingMethodInvokerHelper.java#L541

Reply:
@artembilan thank you for your pro tip.